Order model

DavideBrambilla98 opened the pull request:
Ho aggiunto la classe per il model dell'ordine che permette di:
- Creare nuovi ordini
- Modificare ordini già esistenti
- Eliminare ordini

Questo modello gestisce sia i record dell'ordine sia i record di prodottiOrdini del DB.

DavideBrambilla98 committed:
Aggiunta classe OrderModel

Implementata la classe OrderModel nella quale vengono inseriti tutti i parametri dell'ordine (tranne i prodotti e le relative quantità) nel modello.

DavideBrambilla98 committed:
Aggiunta classe ProductOrderModel

Classe che uso per associare i prodottti agli ordini
sono ancora da implementare alcuni metodi!

DavideBrambilla98 committed:
Mod non presa nel commit precedente

DavideBrambilla98 committed:
Merge remote-tracking branch 'origin/GeneralDevelopment' into OrderModelBranch

DavideBrambilla98 committed:
Modifiche al modello per la gestione degli ordini

- Aggiunti metodi toString agli enumeratori relativi agli ordini
- Cancellata la classe ProductOrderModel poichè gestisco il record della tabella
   ProdottiOrdini direttamente nel modello "OrderModel"
- Aggiunti alcuni dei metodi per la gestione dei record "ProdottiOrdini"

DavideBrambilla98 committed:
OrderModel Class modified

- Aggiunti metodi che permettono di gestire la "listaProdottiOrdiniRecord" e i record
  che contiene.
- Modificata la visibilità da private a package di alcuni metodi della classe
  ProductModel, poichè mi sono serviti nella classe OrderModel

DavideBrambilla98 committed:
Aggiornamento classe OrderModel e StatoOrdine

- Aggiunto metodo che permette di modificare lo stato dell'ordine
- Aggiunta JavaDoc dei vari metodi contenuti nella classe OrderModel
- Corretto metodo getListaProdottiFromDb
- Modificato enumeratore StatoOrdine

DavideBrambilla98 committed:
Update OrderModel.java

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/data/Ordine.java:
```diff
@@ -146,4 +148,11 @@ public static void main(String[] args) {
 		System.out.println(o.toString());
 	}
+
+	public void inserisciProdotto(Prodotto o, int q) throws IOException {
```
Questo metodo lo implementerei nel _modello_ e non nella classe Prodotto, perché è lui che si interfaccia al DB

DavideBrambilla98 replied:
Ho cancellato il metodo dalla classe ordine e l'ho implementato nella classe OrderModel:

	public void inserisciProdotto(Prodotto p, int q)
			throws SQLIntegrityConstraintViolationException {

		getListaProdottiFromDb(this.ordine.getId());

		if (p.isValid() && q > 0)
			createProdottoOrdineRecord();
		else
			throw new SQLIntegrityConstraintViolationException(
					"Ordine #" + p.getId() + " prodotto non valido o qta < 0");
	}

DavideBrambilla98 replied:
Rimosso il metodo dalla classe Ordine e implementato nella classe OrderModel:

	public void inserisciProdotto(Prodotto p, int q)
			throws SQLIntegrityConstraintViolationException {

		getListaProdottiFromDb(this.ordine.getId());

		if (p.isValid() && q > 0) {
			createProdottoOrdineRecord(p, q);
		} else
			throw new SQLIntegrityConstraintViolationException(
					"Ordine #" + p.getId() + " prodotto non valido o qta < 0");
	}

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+	/**
+	 * Crea un nuovo record dell'ordine usando l'ordine del modello
+	 */
+	public void createOrdineRecord()
```
lo metterei private perché gestisce i record

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+	/**
+	 * crea un nuovo record di prodottoOrdine
+	 */
+	public void createProdottoOrdineRecord() {
```
anche questo private
```suggestion
	private void createProdottoOrdineRecord() {
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+	 * @param p   prodotto
+	 * @param qta nuova quantità da inserire nel record
+	 */
+	public void updateQtaProdottoOrdine(Ordine o, Prodotto p, int qta)
```
essendo d'istanza l'ordine non server passarlo come parametro

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+	 * @param o ordine
+	 * @param p prodotto
+	 */
+	public void deleteProdottoOrdine(Ordine o, Prodotto p)
```
anche qua

DavideBrambilla98 replied:
Modificato lasciando solo prodotto da passare al metodo poichè serve a scegliere dal DB il record corretto da eliminare.

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+			this.ordine.setTipo(o.getTipo());
+			this.ordine.setStato(o.getStato());
+			this.ordine.setDataEmissione(o.getDataEmissione());
+			this.ordine.setDataCompletamento(o.getDataCompletamento());
```
Manca di recuperare anche i suoi prodotti con le qta e impostarle con ordine.setProdotti() con la hashmap!

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+	 * @param idP identificativo del prodotto
+	 * @return ritorna vero se esiste
+	 */
+	public boolean productOrderIsSavedInDb(int idO, int idP) {
```
same

DavideBrambilla98 replied:
non ho capito che modifica devo fare.
Intendi che dovrei controllare che nel DB sia stata inserita anche la quantità per dire se esiste o no il record?

JohnnyLAmpAz replied:
No, semplicemente come sopra è inutile dover passare anche idOrdine perché hai già a disposizione l'attributo d'istanza dell'ordine gestito dall'istanza del modello. Non `idO` ma `this.ordine.getId()`.

DavideBrambilla98 replied:
Ah ok, sistemato

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -0,0 +1,370 @@
+		Date dataEmissione = new SimpleDateFormat("dd/MM/yyy")
+				.parse(r.getDataem());
+		Date dataCompletamento = new SimpleDateFormat("dd/MM/yyy")
+				.parse(r.getDataco());
```
Anche qui bisogna tenere conto che non inserisce nell'oggetto ordine la hashmap prodotti-qta! Vista l'utilità del metodo si può pensare di tenerlo così o (meglio) di rinominarlo _ordineFromOrdineRecord_ e di chiarire nel javadoc

DavideBrambilla98 replied:
Il problema è che ordine record non gestisce i prodotti, infatti anche quando viene creato un nuovo record di un ordine, non viene passata la HashMap dei prodotti.

Ecco il costruttore di un record dell'ordine:

    public OrdineRecord(Integer id, String tipo, String stato, String dataem, String dataco) {
        super(Ordine.ORDINE);

        setId(id);
        setTipo(tipo);
        setStato(stato);
        setDataem(dataem);
        setDataco(dataco);
        resetChangedOnNotNull();
    }

Per gestire i prodotti e le relative quantità si utilizzano i ProdottiordiniRecord.

JohnnyLAmpAz replied:
Vero ma si possono settare con

https://github.com/JohnnyLAmpAz/smartmag/blob/dd8c1d03d2ed4d0b87ed770663810826caebe5f8/proj_eclipse/src/main/java/smartmag/data/Ordine.java#L80-L82

DavideBrambilla98 replied:
Ma questo metodo non aggiorna gli oggetti della classe ordine, aggiorna i record di ordine.
Dato che nei record di ordine non sono presenti i prodotti non capisco come potrei aggiornarli.

DavideBrambilla98 replied:
Se intendi aggiornare gli oggetti della classe ordine partendo da un record invece, dovrei fare un nuovo metodo getOrderFromDb o qualcosa di simile.
Ma questo metodo deve accedere sia a OrderRecord che a ProductOrderRecord per avere tutti gli attributi necessari per aggiornare l'oggetto ordine.

JohnnyLAmpAz replied:
Nono, è dell'istanza Ordine! ordine.setProdotti()

DavideBrambilla98 replied:
Ma aggiorna i prodotti dell'istanza dell'ordine il metodo setProdotti()
Non lo posso usare per aggiornare il record di un ordine

JohnnyLAmpAz replied:
> Restituisce un ordine partendo dal suo record

Scritto così pensavo dovesse restituire l'intero oggetto Ordine (**prodotti compresi**) dai suoi record nelle due tabelle. Se invece intendi ritornare un oggetto ordine senza lista prodotti allora va bene così. Diciamo che in effetti dovresti avere anche i recordProdottiOrdine che
- o passi come parametri
- o ti ricavi in altro modo

DavideBrambilla98 replied:
ah ma pensavo ti riferissi alla classe copyOrdineIntoRecord, invece intendi ordineFromOrdineRecord

DavideBrambilla98 replied:
Si si hai ragione, meglio recuperare anche i prodotti dell'ordine.
Per farlo al posto che usare OrdineRecord, uso ProdottiordiniRecord e dovrei riuscire.

JohnnyLAmpAz reviewed (changes requested):
Guarda i commenti che ho messo. Per il resto, LGTM 💪.

DavideBrambilla98 committed:
Modifiche classe OrderModel

Effettuate le modifiche suggerite nei commenti della PR#9

DavideBrambilla98 committed:
Update OrderModel.java

Aggiornato metodo ordineFromOrdineRecord:
adesso il metodo ritorna l'ordine completo (con lista prodotti e qta)

JohnnyLAmpAz reviewed (approved):
Goooood 👍
